File: java-testng-selenium/src/test/java/selfhealinglocators/TestBase.java
package selfhealinglocators;

import org.testng.annotations.DataProvider;

public class TestBase 
{
    String site1 = "https://juanfonsecagl.github.io/SelfHealingLocatorsDemo/demoSite/form-button-label.html";
    String site2 = "https://juanfonsecagl.github.io/SelfHealingLocatorsDemo/demoSite/form-button-label2.html";
    String site3 = "https://juanfonsecagl.github.io/SelfHealingLocatorsDemo/demoSite/form-button-label3.html";

    @DataProvider(name = "testData")
    public Object[][] testData() {
        return new Object[][] 
        {
            // TestSuite #1: different button implementation (scenario proposed by Test Rigor)
            { site1,    "//input[@id='messageNew']",                "//button[@id='changer']",  "correct" },    // TC1
            { site2,    "//input[@id='messageNew']",                "//a[@id='pusher']",        "correct" },    // TC2
            { site1,    "//input[@id='messageNew']",                "//a[@id='pusher']",        "broken" },     // TC3
            { site2,    "//input[@id='messageNew']",                "//button[@id='changer']",  "broken" },     // TC4

            // TestSuite #2: different placeholder for input (scenario proposed by Healenium)
            { site1,   "//input[@placeholder='Message']",          "//button[@id='changer']",   "correct" },    // TC5
            { site3,   "//input[@placeholder='Enter some text']",  "//button[@id='changer']",   "correct" },    // TC6
            { site1,   "//input[@placeholder='Enter some text']",  "//button[@id='changer']",   "broken" },     // TC7
            { site3,   "//input[@placeholder='Message']",          "//button[@id='changer']",   "broken" }      // TC8
        };
    }
}

File: dotnet-nunit-selenium/Tests/TestBase.cs
using OpenQA.Selenium;

public class TestBase
{
    protected bool _testPassed;
    protected string _reportFilePath = Path.Combine(Directory.GetCurrentDirectory(),"TestResultsSummary.txt");
    
    const string _baseSite = "https://juanfonsecagl.github.io/SelfHealingLocatorsDemo/demo-site/demo-site1";
    const string _site1 = $"{_baseSite}/form-button-label.html";
    const string _site2 = $"{_baseSite}/form-button-label2.html";
    const string _site3 = $"{_baseSite}/form-button-label3.html";

    public static object[] TestData =
    {
        // TestSuite #1: different button implementation (scenario proposed by TestRigor)
        new object[] { "TC1", _site1,   By.XPath("//input[@id='messageNew']"),                By.XPath("//button[@id='changer']"),  "correct" }, 
        new object[] { "TC2", _site2,   By.XPath("//input[@id='messageNew']"),                By.XPath("//a[@id='pusher']"),        "correct" }, 
        new object[] { "TC3", _site1,   By.XPath("//input[@id='messageNew']"),                By.XPath("//a[@id='pusher']"),        "broken" }, 
        new object[] { "TC4", _site2,   By.XPath("//input[@id='messageNew']"),                By.XPath("//button[@id='changer']"),  "broken" }, 

        // TestSuite #2: different placeholder for input (scenario proposed by Healenium)
        new object[] { "TC5", _site1,   By.XPath("//input[@placeholder='Message']"),          By.XPath("//button[@id='changer']"),   "correct" },  
        new object[] { "TC6", _site3,   By.XPath("//input[@placeholder='Enter some text']"),  By.XPath("//button[@id='changer']"),   "correct" }, 
        new object[] { "TC7", _site1,   By.XPath("//input[@placeholder='Enter some text']"),  By.XPath("//button[@id='changer']"),   "broken" },
        new object[] { "TC8", _site3,   By.XPath("//input[@placeholder='Message']"),          By.XPath("//button[@id='changer']"),   "broken" } 
    };
}